### PR TITLE
Add unit test for edges API endpoint

### LINKED ISSUE #3052: Add unit test for `edges` API endpoint
Now that the API endpoint for `linkerd edges` is in good shape, we need a unit test for this endpoint, to join the other tests in `controller/api/public`.

### PR DESCRIPTION
**Pull request by scottcarol**
Fixes #3052.

This PR adds a unit test for the `edges` API endpoint. 
* `linkerd edges` does not query the K8s API, it just returns Prometheus data, so we do not need k8s configs.
* Because it is important to test specifying a namespace (i.e. `linkerd edges deploy -n emojivoto`), `GenEdgesResponse` now accepts an `edgeRowNamespace` string. This affects both the CLI and API unit test for edges.  
* To maintain consistent test results we add sorting both to the returned edge rows in `api/public/edges.go` and the expected edge rows in `/api/public/test_helper.go`.
* Because I reuse `expectedStatRPC` and `newMockGrpcServer` I generate a single `MockPromResponse` that contains both SRC and DST responses, although `linkerd edges` actually makes 2 separate Prometheus queries. This seemed more efficient than creating multiple new functions based on `newMockGrpcServer` but I'm open to feedback on that.

### CODE REVIEW

**Comment by ihcsim** [comment]
Looks like `lint` got a bit upset:
```
controller/api/public/edges_test.go:24:27: `genInboundPromSample` - `resourceType` always receives `"deployment"` (unparam)
func genInboundPromSample(resourceType, resourceNamespace, resourceName, clientID string) *model.Sample {
                          ^
controller/api/public/edges_test.go:38:28: `genOutboundPromSample` - `resourceType` always receives `"deployment"` (unparam)
func genOutboundPromSample(resourceType, resourceNamespace, resourceName, resourceNameDst, resourceNamespaceDst, serverID string) *model.Sample {
                           ^
controller/api/public/test_helper.go:387: File is not `gofmt`-ed with `-s` (gofmt)
	&mockEdgeRow{
level=info msg="[runner] worker.1 took 42.361555541s with stages: megacheck: 35.717608594s, goimports: 6.323791104s, varcheck: 126.486588ms, deadcode: 83.774996ms, unconvert: 63.683993ms, goconst: 28.490809ms, nakedret: 16.734947ms, typecheck: 7.204µs"
level=info msg="[runner] worker.2 took 42.681053971s with stages: bodyclose: 13.277658661s, golint: 12.947320605s, govet: 12.379436304s, gofmt: 1.561237973s, ineffassign: 912.553026ms, unparam: 784.49775ms, misspell: 757.513443ms, scopelint: 50.455076ms, depguard: 20.605µs"
level=info msg="[runner] Workers idle times: #1: 326.907626ms"
level=info msg="[runner] Issues before processing: 1889, after processing: 3"
level=info msg="[runner] processing took 113.672893ms with stages: skip_dirs: 82.223436ms, autogenerated_exclude: 13.960747ms, cgo: 8.076981ms, path_prettifier: 3.987203ms, source_code: 2.731662ms, identifier_marker: 1.265801ms, filename_unadjuster: 674.82µs, nolint: 596.211µs, max_same_issues: 53.421µs, skip_files: 36.603µs, path_shortener: 29.089µs, uniq_by_line: 13.469µs, max_per_file_from_linter: 7.664µs, max_from_linter: 6.801µs, exclude-rules: 3.207µs, diff: 3.039µs, exclude: 2.739µs"
level=info msg="File cache stats: 238 entries of total size 1.5MiB"
level=info msg="Memory: 90 samples, avg is 1548.0MB, max is 6736.3MB"
level=info msg="Execution took 53.89452089s"
The command "./bin/lint --verbose" exited with 1.
```

File: controller/api/public/edges_test.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+		if !proto.Equal(exp.expectedResponse.GetOk(), rsp.GetOk()) {
+			t.Fatalf("Expected edgesOkResp: %+v\n Got: %+v", &exp.expectedResponse, rsp)
+		}
+	}
+}
+
+func TestEdges(t *testing.T) {
+	mockPromResponse := model.Vector{
+		genInboundPromSample("deployment", "emojivoto", "emoji", "web.emojivoto.serviceaccount.identity.linkerd.cluster.local"),
```
Comment (ihcsim) [inline controller/api/public/edges_test.go]: For my own understanding, this represents mock metrics of inbound traffic originated from `web`, as seen by the `emoji` proxy, right?

File: controller/api/public/edges_test.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		genInboundPromSample("deployment", "emojivoto", "voting", "web.emojivoto.serviceaccount.identity.linkerd.cluster.local"),
+		genInboundPromSample("deployment", "emojivoto", "web", "default.emojivoto.serviceaccount.identity.linkerd.cluster.local"),
+		genInboundPromSample("deployment", "linkerd", "linkerd-prometheus", "linkerd-controller.linkerd.identity.linkerd.cluster.local"),
+
+		genOutboundPromSample("deployment", "emojivoto", "web", "emoji", "emojivoto", "emoji.emojivoto.serviceaccount.identity.linkerd.cluster.local"),
```
Comment (ihcsim) [inline controller/api/public/edges_test.go]: For my own understanding, this represents mock metrics of outbound traffic, as seen by the `web` proxy, with `emoji` being the destination?

File: controller/api/public/edges_test.go (hunk 1 above)
Comment (scottcarol) [reply controller/api/public/edges_test.go]: That's correct. Line 99 generates these mock Prometheus metrics: 
`{client_id="web.emojivoto.serviceaccount.identity.linkerd.cluster.local", deployment="emoji", namespace="emojivoto"} => 123 @[0.456]`
which would be one `Sample` in a `Vector` returned by Prometheus in response to a query for all **inbound** requests to deployments. We match the `inbound` metrics to the `outbound` metrics in order to construct a complete edge.

File: controller/api/public/edges_test.go (hunk 2 above)
Comment (scottcarol) [reply controller/api/public/edges_test.go]: That's right.

File: cli/cmd/edges_test.go [hunk 3]
```diff
@@ -169,7 +169,7 @@ func TestEdges(t *testing.T) {
 
 func testEdgesCall(exp edgesParamsExp, t *testing.T) {
 	mockClient := &public.MockAPIClient{}
-	response := public.GenEdgesResponse(exp.resourceType, exp.resSrc, exp.resDst, exp.resSrcNamespace, exp.resDstNamespace, exp.resClient, exp.resServer, exp.resMsg)
+	response := public.GenEdgesResponse(exp.resourceType, "all")
```
[the remark this answers is not in the document]
Comment (scottcarol) [reply cli/cmd/edges_test.go]: Good catch thanks @ihcsim!

File: controller/api/public/test_helper.go [hunk 4]
```diff
@@ -371,29 +371,100 @@ func GenStatTsResponse(resName, resType string, resNs []string, basicStats bool,
...
+	if edgeRowNamespace == "linkerd" {
+		edgeRows = linkerdEdgeRows
+	} else if edgeRowNamespace == "all" {
+		// combine emojivotoEdgeRows and linkerdEdgeRows
+		edgeRows = append(edgeRows, linkerdEdgeRows...)
+	}
+
 	edges := []*pb.Edge{}
-	for i := range resSrc {
+	for _, i := range edgeRows {
```
Comment (siggy) [inline controller/api/public/test_helper.go]: nit: `i` is typically for the index var, the second one would be something like `edge` or `row` or `edgeRow`.

**Pull request merged**